src/jira/util/jwt.ts

```diff
@@ -211,41 +195,136 @@ const isStagingTenant = (req: Request): boolean => {
 	try {
 		const token = extractJwtFromRequest(req);
-		if (!token) {
-			req.log.info("JWT Verification Failed, no token present");
-			sendError(res, 401, "Unauthorized");
-			return;
-		}
+		await validateAsymmetricJwtToken(getJWTRequest(req), token, isStagingTenant(req));
+		req.log.info("JWT Token Verified Successfully!");
+		next();
+	} catch (err) {
+		req.log.info(err, "Could not validate JWT token");
+		res.status(401).json({
+			message: "Unauthorized"
+		});
+	}
+};
 
-		const publicKey = await queryAtlassianConnectPublicKey(getKeyId(token), isStagingTenant(req));
-		const unverifiedClaims = decodeAsymmetricToken(token, publicKey, true);
+export const validateAsymmetricJwtToken = async (request: JWTRequest, token?: string, isStaginTenant = false) => {
 
-		const issuer = unverifiedClaims.iss;
-		if (!issuer) {
-			req.log.info("JWT Verification Failed, no issuer present");
-			sendError(res, 401, "JWT claim did not contain the issuer (iss) claim");
-			return;
-		}
+	if (!token) {
+		throw "JWT Verification Failed, no token present";
+	}
 
-		if (isEmpty(unverifiedClaims.aud) ||
-			!unverifiedClaims.aud[0] ||
-			!includes(ALLOWED_BASE_URLS, unverifiedClaims.aud[0].replace(/\/$/, ""))) {
-			req.log.info("JWT Verification Failed, no correct audience present");
-			sendError(res, 401, "WT claim did not contain the correct audience (aud) claim");
-			return;
-		}
+	const publicKey = await queryAtlassianConnectPublicKey(getKeyId(token), isStaginTenant);
+	const unverifiedClaims = decodeAsymmetricToken(token, publicKey, true);
 
-		const verifiedClaims = decodeAsymmetricToken(token, publicKey, false);
+	const issuer = unverifiedClaims.iss;
+	if (!issuer) {
+		throw "JWT claim did not contain the issuer (iss) claim";
+	}
 
-		if (verifyJwtClaimsAndSetResponseCodeOnError(verifiedClaims, TokenType.normal, req, res)) {
-			req.log.info("JWT Token Verified Successfully!");
-			next();
-		}
-	} catch (e) {
-		req.log.warn({ ...e }, "Error while validating JWT token");
-		sendError(res, 401, "Unauthorized");
+	if (isEmpty(unverifiedClaims.aud) ||
+		!unverifiedClaims.aud[0] ||
+		!includes(ALLOWED_BASE_URLS, unverifiedClaims.aud[0].replace(/\/$/, ""))) {
+		throw "JWT claim did not contain the correct audience (aud) claim";
+	}
+
+	const verifiedClaims = decodeAsymmetricToken(token, publicKey, false);
+
+	validateJwtClaims(verifiedClaims, TokenType.normal, request);
+};
+
+export interface JWTRequest extends UrlWithParsedQuery {
+	method: string;
+	body?: any;
+}
+
+export const getJWTRequest = (req: Request): JWTRequest => ({
+	...url.parse(req.originalUrl || req.url, true),
+	method: req.method,
+	body: req.body
+});
+
+const CANONICAL_QUERY_SEPARATOR = "&";
+
+enum HASH_ALGORITHM {
+	HS256 = "sha256",
+	HS384 = "sha384",
+	HS512 = "sha512",
+	RS256 = "RSA-SHA256"
+}
+
+export const createQueryStringHash = (req: JWTRequest, checkBodyForParams?: boolean): string => {
```

Comment:
/nit
I'd suggest have this in a separate file and mention the origin file it comes from. So that next dev will easier to pick it up for figure out which team to looks for (if say they change the way `createCanonicalRequest` works in jira)